Convert build system to Gradle

FuegoFro opened the pull request:
This roughly follows the [IntelliJ plugin template](https://github.com/JetBrains/intellij-platform-plugin-template/blob/master/build.gradle.kts) to convert the project to be built with Gradle. This should allow folks to compile the project more easily and reliably, and is also the recommended way to build projects moving forward.

As part of this, and following [best practices (1.3)](https://github.com/JetBrains/Grammar-Kit/blob/master/HOWTO.md#i-general-notes) and moves the generated sources into a separate source root, in this case the `gen` folder.

Note that the main difference between this project and the default/recommended/template project is the source layout. Namely, they expect `src` dir that has `java`, `resources`, etc in it, whereas our source code is directly in the `src` directory. If folks here are open to it, I think moving the code could make it easier to work with the plugin moving forward.

FuegoFro committed:
Convert build system to Gradle

This roughly follows the [IntelliJ plugin template](https://github.com/JetBrains/intellij-platform-plugin-template/blob/master/build.gradle.kts) to convert the project to be built with Gradle. This should allow folks to compile the project more easily and reliably, and is also the recommended way to build projects moving forward.

As part of this, and following [best practices (1.3)](https://github.com/JetBrains/Grammar-Kit/blob/master/HOWTO.md#i-general-notes) and moves the generated sources into a separate source root, in this case the `gen` folder.

Note that the main difference between this project and the default/recommended/template project is the source layout. Namely, they expect `src` dir that has `java`, `resources`, etc in it, whereas our source code is directly in the `src` directory. If folks here are open to it, I think moving the code could make it easier to work with the plugin moving forward.

jim-hart commented:
Hi @FuegoFro, welcome!

@valery-labuzhsky  can you review this?

@slideclimb does this affect your work on #66?

slideclimb commented:
@jim-hart No, I suppose all will be fine when I merge this into the #66 branch :) I had a quick look at this PR and it looks good to me. I do agree with @FuegoFro that moving the code to follow conventions with a `java` and `resource` folder does make sense at this point. Even though it might make merging with #66 a bit more tedious, but I'm sure I'll be able to handle it.

Thanks for the heads-up!

valery-labuzhsky commented:
Hi @FuegoFro 
Thank for the contribution! 

Everything work fine. I have a couple of notes to compile instructions though.

1. I don't see gradlew script, so step #2 results in "bash: ./gradlew: No such file or directory"
I managed to run it using gradle menu in idea. Any steps missing here?

2. Gradle project broke my previous run configuration. Mainly because module no longer detected as a plugin module. I managed to run it from gradle runIde task. Is it supposed to be this way? If so, worth changing #3 to something like run "gradle runIde task" 

As of folder structure - if changing them makes someones life easier - let's do it.

FuegoFro committed:
Add Gradle wrapper files, make readme a bit clearer.

FuegoFro commented:
Hi! Thanks so much for the quick reviews!

@valery-labuzhsky those are good catches!
1. I think a poor rebase on my part from #66 lost the wrapper files, I've added them back in 🙂  They're generated with the [`gradle wrapper`](https://docs.gradle.org/current/dsl/org.gradle.api.tasks.wrapper.Wrapper.html) task and are intended to be checked in according to the docs.
2. To be perfectly honest I'm not 100% certain how the run configuration worked before (it doesn't look like anything in the `.idea` folder was/is checked in?), but yeah that will certainly work 🙂 It's worth noting that that task starts up a sandboxed copy of the IDE and I'm not sure how/whether things like settings for it get preserved. In any case I've updated the `README.me` to have a bit more detail here 🙂

FuegoFro commented:
I'll put moving the files in a separate PR since it's kinda a different concern (unless y'all would rather I put it here, in which case I'm happy to do that too 🙂). I... don't know that it's possible to set up dependent PRs *between repos* with GitHub (eg, I can't use `idea_gradle` my branch as the base when setting up a PR into this repo since that branch only exists in my repo) so I'll hold off on submitting the move files PR until this one is merged, but it looks like this: https://github.com/FuegoFro/EditorTools/compare/idea_gradle...FuegoFro:idea_move_src_files?expand=1

valery-labuzhsky commented:
@FuegoFro now readme is perfect! Thank you.

One last problem is that gradlew is not executable. Here is the command I've found on [stackoverflow](https://stackoverflow.com/questions/21691202/how-to-create-file-execute-mode-permissions-in-git-on-windows):
`git update-index --chmod=+x gradlew`  
Worked for me even on linux.

Separate PR is perfectly fine.
JFYI, you can merge _idea_move_src_files_ to _idea_gradle_. That way it should in theory appear in this PR. I cannot say I've ever done it on github though.

FuegoFro committed:
Make gradlew executable

FuegoFro commented:
lol, comedy of errors over here 🤦 It's executable now. I'll chalk it up to not being used to working on Windows heh.

valery-labuzhsky commented:
@jim-hart please merge.

jim-hart commented:
@FuegoFro Thank you for addressing all issues!
@slideclimb  No Problem!
@valery-labuzhsky  Thank you for the Review!

FuegoFro commented:
Thank you for the merge! 😀
